### PR TITLE
Search

### PR DESCRIPTION
**Pull request by dcunhas**
I added a search bar to the URL and Substitution areas. Because I'm not a web designer, the added UI looks a little rough but it is functional. Please make some suggestions on changes necessary before this can be merged. This should resolve issue #134.

### CODE REVIEW

**Comment by Woundorf** [comment]
I will need a few days (or weeks) to review this appropriately and test it, but on a superficial level I see that there is inconsistent indentation, mixing spaces and tabs. Please convert them all to the appropriate number of spaces.

**Comment by dcunhas** [comment]
Sorry about that, didn't realize my editor was not configured correctly. The space issue should be fixed now. Take your time reviewing, I'm in no rush to have it merged.

**Comment by Woundorf** [comment]
Hi, I have now had some time to review it in detail and test it. I have found a few issues that would need to be resolved. I will also add a few comments to specific lines for other minor things.

#### 1. Sometimes the last empty row that allows to add new elements can be duplicated (or triplicated, etc.)

Steps to reproduce:

1. Edit a group with some entries (it doesn't matter if URLs or substitutions, it happens on both places).
2. Write something in the search box allowing some element to be visible.
3. Start to edit one of the visible items.
4. Clear the search box.

With these steps a new empty row is added at the end. It is added by this fragment in groupeditor.js (this one for urls and a similar one for substitutions):

```javascript
      let isLast = isLastRow(params);
      let isEmpty = params.value === "";

      if (!isLast && isEmpty) {
        params.api.updateRowData({ remove: [params.data] });
      }
      else if (isLast && !isEmpty) {
        params.api.updateRowData({ add: [{ url: "" }] });
      }
```

The conditions could be adjusted to take into account the search (or maybe there's a better solution).

#### 2. Move actions in the substitutions tab don't behave well when a filter is active

Unless you can think of a way of making them work, they could be disabled when the filter is active.

File: options/groupeditor.js [hunk 1]
```diff
@@ -259,22 +259,36 @@ var groupEditor = (() => {
...
     clear() {
       editor.clearUrls();
+    },
+    onSearchInput(event) {
+        let api = urlsGridOptions.api;
+        api.setQuickFilter(document.getElementById('urlsSearchBar').value);
+    },
+    onClickClearSearch(event) {
+        document.getElementById('urlsSearchBar').value = '';
+        urlsEventListeners.onSearchInput(event);
```
Comment (Woundorf) [inline options/groupeditor.js:269]: Please adjust indentation to two spaces instead of four.

File: options/groupeditor.js [hunk 2]
```diff
@@ -259,22 +259,36 @@ var groupEditor = (() => {
...
     }
+
```
Comment (Woundorf) [inline options/groupeditor.js:271]: Please delete this line.

File: options/groupeditor.js [hunk 3]
```diff
@@ -259,22 +259,36 @@ var groupEditor = (() => {
...
 
   function removeUrlsEventListeners() {
     document.getElementById("urlsGrid").removeEventListener("keyup", urlsEventListeners.onKeyUp);
     document.getElementById("urlsGrid").removeEventListener("keydown", urlsEventListeners.onKeyDown, true);
     document.getElementById("clearUrlsButton").removeEventListener("click", urlsEventListeners.clear);
+    document.getElementById("urlsSearchBar").removeEventListener("input", urlsEventListeners.onSearchInput);
+    document.getElementById("urlsSearchClear").removeEventListener("click", urlsEventListeners.onClickClearSearch);
   }
 
+
```
Comment (Woundorf) [inline options/groupeditor.js:291]: Please delete this line.

File: options/groupeditor.js [hunk 4]
```diff
@@ -348,6 +362,14 @@ var groupEditor = (() => {
...
       if (!api.getSelectedNodes()) return null;
       else return api.getSelectedNodes()[0];
+    },
+    onSearchInput(event) {
+        let api = substitutionsGridOptions.api;
+        api.setQuickFilter(document.getElementById('substitutionsSearchBar').value);
+    },
+    onClickClearSearch(event) {
+        document.getElementById('substitutionsSearchBar').value = '';
+        substitutionsEventListeners.onSearchInput(event);
```
Comment (Woundorf) [inline options/groupeditor.js:372]: Please adjust indentation to two spaces instead of four.

File: options/groupeditor.js [hunk 5]
```diff
@@ -359,6 +381,9 @@ var groupEditor = (() => {
     document.getElementById("moveDownSubstitutionButton").addEventListener("click", substitutionsEventListeners.moveDown);
     document.getElementById("moveBottomSubstitutionButton").addEventListener("click", substitutionsEventListeners.moveBottom);
     document.getElementById("clearSubstitutionsButton").addEventListener("click", substitutionsEventListeners.clear);
+    document.getElementById("substitutionsSearchBar").addEventListener("input", substitutionsEventListeners.onSearchInput);
+    document.getElementById("substitutionsSearchClear").addEventListener("click", substitutionsEventListeners.onClickClearSearch);
+
```
Comment (Woundorf) [inline options/groupeditor.js:386]: Please delete this empty line.

File: options/groupeditor.js [hunk 6]
```diff
@@ -369,6 +394,9 @@ var groupEditor = (() => {
     document.getElementById("moveDownSubstitutionButton").removeEventListener("click", substitutionsEventListeners.moveDown);
     document.getElementById("moveBottomSubstitutionButton").removeEventListener("click", substitutionsEventListeners.moveBottom);
     document.getElementById("clearSubstitutionsButton").removeEventListener("click", substitutionsEventListeners.clear);
+    document.getElementById("substitutionsSearchBar").removeEventListener("input", substitutionsEventListeners.onSearchInput);
+    document.getElementById("substitutionsSearchClear").removeEventListener("click", substitutionsEventListeners.onClickClearSearch);
+
```
Comment (Woundorf) [inline options/groupeditor.js:399]: Please delete this empty line.

File: options/groupeditor.js [hunk 7]
```diff
@@ -446,10 +474,14 @@ var groupEditor = (() => {
 
     isValidGroup() {
       let error = false;
+      let api = substitutionsGridOptions.api;
+      api.setQuickFilter(''); //Disable filter temporarly for validation
```
Comment (Woundorf) [inline options/groupeditor.js:478]: Typo in the comment.

File: options/options.html [hunk 8]
```diff
@@ -196,6 +196,12 @@ <h4 class="i18n">options.otherSettings</h4>
             </ul>
             <div class="tab-content">
               <div id="urlsContent" class="tab-pane show active" role="tabpanel">
+                 <div class="input-group mb-3">
+                    <input  id="urlsSearchBar" type="search" Placeholder="Search" class="form-control searchBar"/>
+                    <span id="urlsSearchClear" class="form-control-feedback form-control-clear closeIcon">
+                        X
+                    </span>
```
Comment (Woundorf) [inline options/options.html:203]: Better use an icon like in other places. Example:

```html
<i class="fas fa-backspace"></i>
```

Available icons that could be suitable: https://fontawesome.com/v5.9/icons?d=gallery&p=2&q=delete&m=free. I think that "backspace" or "eraser" could be the more appropriate.

With this change I think you will be able to simplify the css above too.

File: options/options.html [hunk 9]
```diff
@@ -207,6 +213,12 @@ <h4 class="i18n">options.otherSettings</h4>
                 </div>
               </div>
               <div id="substitutionsContent" class="tab-pane" role="tabpanel">
+                <div class="input-group mb-3">
+                    <input id="substitutionsSearchBar" type="search" Placeholder="Search" class="form-control searchBar"/>
+                    <span id="substitutionsSearchClear" class="form-control-feedback form-control-clear closeIcon">
+                        X
+                    </span>
```
Comment (Woundorf) [inline options/options.html:220]: The same icon as above.

File: options/options.html [hunk 10]
```diff
@@ -241,7 +253,7 @@ <h4 class="i18n">options.otherSettings</h4>
           </div>
           <div class="modal-footer">
             <button id="groupOk" class="btn btn-primary i18n" data-action="ok">options.ok</button>
-            <button class="btn btn-secondary i18n" data-dismiss="modal">options.cancel</button>
+            <button id="grouCancel" class="btn btn-secondary i18n" data-action="cancel">options.cancel</button>
```
Comment (Woundorf) [inline options/options.html:256]: Typo in the id.

File: options/options.js [hunk 11]
```diff
@@ -203,6 +204,7 @@ function addEventListeners() {
   });
   $('#groupOk').click(saveGroup);
   $('#groupApply').click(saveGroup);
+  $('#grouCancel').click(closeGroupEditor);
```
Comment (Woundorf) [inline options/options.js:207]: Same typo in the id as above.

**Review by Woundorf — COMMENTED** [review]
Please address also these minor issues. In general it's looking good.

**Comment by dcunhas** [comment]
I think I fixed everything you mentioned. Notably for the issue with the movement buttons, I just disabled them while searching. I can conceive of (rare) situations where one might want to move an entry's position while searching, but the logic and methodology of that seems nontrivial to me, so that could be considered a possible future feature.

File: options/groupeditor.js [hunk 12]
```diff
@@ -464,8 +520,21 @@ var groupEditor = (() => {
...
+      document.getElementById("substitutionsSearchClear").dispatchEvent(clearSearchEvent);
+    },
+
+    currentlySearchingUrls() {
+      return document.getElementById('urlsSearchBar').value != '';
+    },
+
+    currentlySearchingSubstitutions() {
+      return document.getElementById('substitutionsSearchBar').value != '';
+    },
```
Comment (Woundorf) [inline options/groupeditor.js:537]: Please remove the comma and add blank line after the curly bracket.

File: options/options.html [hunk 13]
```diff
@@ -196,6 +196,10 @@ <h4 class="i18n">options.otherSettings</h4>
             </ul>
             <div class="tab-content">
               <div id="urlsContent" class="tab-pane show active" role="tabpanel">
+                 <div class="input-group mb-3">
+                    <input  id="urlsSearchBar" type="search" Placeholder="Search" class="form-control searchBar"/>
+                    <i id="urlsSearchClear" class="fas fa-times-circle closeIcon"></i>
```
Comment (Woundorf) [inline options/options.html:201]: Adjust indentation to 2 spaces instead of 3.

File: options/options.js [hunk 14]
```diff
@@ -227,11 +229,16 @@ function removeEventListeners() {
...
+  groupEditor.resetSearch();
 
   if (event.relatedTarget == document.getElementById('addGroup')) {
     groupEditor.clear();
   }
 }
+function closeGroupEditor(event) {
+    groupEditor.resetSearch();
+    $('#groupEditorModal').modal('hide');
+}
```
Comment (Woundorf) [inline options/options.js:241]: Add a blank line before the function and adjust indentation to 2 spaces instead of 4.

**Review by Woundorf — COMMENTED** [review]
Now it seems to work properly, I have not detected any other functional issue. I have added a few new comments about spacing issues. Sorry to bother you again about that but I like code to be consistent. I would make the changes myself but I can't edit the pull request myself, as far as I know.

When you make these last requested changes I will merge the pull request. Thank you for your contribution.

**Comment by dcunhas** [comment]
I think I fixed everything, let me know if there's anything more.

**Comment by Woundorf** [comment]
Thanks, now everything is ok.

> On another note, have you considered using something like [prettier](https://github.com/prettier/prettier) in a pre-commit hook to automate code style compliance?

I didn't know this particular tool, but also since I don't usually receive contributions I haven't had any need for it.

**Pull request merged**